src/validation.cpp

```diff
@@ -2854,6 +2897,143 @@ bool CChainState::ConnectBlock(const CBlock& block, CValidationState& state, CBl
+                    }
+                    auto vault = cache.GetVault(vaultId);
+                    assert(vault);
+                    auto scheme = cache.GetLoanScheme(vault.val->schemeId);
+                    assert(scheme);
+                    if (scheme->ratio <= collateral->ratio()) {
+                        return true;
+                    }
+                    vault.val->isUnderLiquidation = true;
+                    cache.StoreVault(vaultId, *vault.val);
+                    auto loanTokens = cache.GetLoanTokens(vaultId);
+                    assert(loanTokens);
+                    for (const auto& loan : loanTokens->balances) {
+                        cache.SubLoanToken(vaultId, {loan.first, loan.second});
+                    }
+                    for (const auto& col : collaterals.balances) {
+                        cache.SubVaultCollateral(vaultId, {col.first, col.second});
+                    }
+                    auto batches = CollectAuctionBatches(*collateral, collaterals.balances, loanTokens->balances);
+                    for (auto i = 0u; i < batches.size(); i++) {
+                        cache.StoreAuctionBatch(vaultId, i, batches[i]);
+                    }
+                    cache.StoreAuction(vaultId, pindex->nHeight, CAuctionData{
+                        uint32_t(batches.size()), cache.GetLoanLiquidationPenalty()
+                    });
+                    return true;
+                });
+            }
+            cache.ForEachVaultAuction([&](const AuctionKey& auction, const CAuctionData& data) {
+                if (int(auction.height) != pindex->nHeight) {
+                    return false;
+                }
+                for (uint32_t i = 0; i < data.batchCount; i++) {
+                    auto batch = cache.GetAuctionBatch(auction.vaultId, i);
+                    assert(batch);
+                    if (auto bid = cache.GetAuctionBid(auction.vaultId, i)) {
+                        auto amountToFill = DivideAmounts(bid->second.nValue, COIN + data.liquidationPenalty);
+                        auto amountToBurn = bid->second.nValue - amountToFill;
+                        if (amountToBurn > 0) {
+                            cache.AddBalance(chainparams.GetConsensus().burnAddress, {bid->second.nTokenId, amountToBurn});
+                            CPoolSwapMessage obj;
+                            obj.from = chainparams.GetConsensus().burnAddress;
+                            obj.to = chainparams.GetConsensus().burnAddress;
+                            obj.idTokenFrom = bid->second.nTokenId;
+                            obj.idTokenTo = DCT_ID{0};
+                            obj.amountFrom = amountToBurn;
+                            obj.maxPrice = POOLPRICE_MAX;
+                            auto poolSwap = CPoolSwap(obj, pindex->nHeight);
+                            // swap tokenID -> USD -> DFI
+                            auto token = cache.GetToken("USD");
```

Comment:
In the pink paper, it should swap to DFI and burn.

Reply:
I will do it this week.